File: algorithms/Implement Queue using Stacks/ImplementQueueUsingStack.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2016-11-30 21:13:57
# @Author  : luckcul (tyfdream@gmail.com)
# @Version : 2.7.12

class Queue(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []

    def push(self, x):
        """
        :type x: int
        :rtype: nothing
        """
        self.stack.append(x)

    def pop(self):
        """
        :rtype: nothing
        """
        sta = []
        l = len(self.stack)
        for i in range(l):
        	tmp = self.stack.pop()
        	sta.append(tmp)
        sta.pop()
        for i in range(l-1):
        	tmp = sta.pop()
        	self.stack.append(tmp)

    def peek(self):
        """
        :rtype: int
        """
        sta = []
        l = len(self.stack)
        for i in range(l):
        	tmp = self.stack.pop()
        	sta.append(tmp)
        ret = sta.pop()
        self.stack.append(ret)
        for i in range(l-1):
        	tmp = sta.pop()
        	self.stack.append(tmp)
        return ret

    def empty(self):
        """
        :rtype: bool
        """
        return True if len(self.stack)==0 else False

x = Queue()
x.push(1)
x.push(2)
print x.peek()
x.pop()
print x.peek()
print x.empty()
x.pop()
print x.empty()

File: algorithms/Implement Queue using Stacks/ImplementQueueUsingStack.cpp
/* ***********************************************
 	Author        : luckcul
 	Mail          : tyfdream@gmail.com
 	Created Time  : 2016-11-30 22:21:12
 	Problem       : problem
************************************************ */
#include <cstdio>
#include <cstring>
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
#include <stack>
#include <queue>
#include <map>
#include <set>
using namespace std; 
#define INF 1000000000
//typedef __int64 LL; 

class Queue {
	stack<int> Stack;
public:
    // Push element x to the back of queue.
    void push(int x) {
        Stack.push(x);
    }

    // Removes the element from in front of queue.
    void pop(void) {
        int l = Stack.size();
        stack<int> S;
        for(int i = 0; i < l; i++) {
        	int tmp = Stack.top();
        	Stack.pop();
        	S.push(tmp);
        }
        S.pop();
        for(int i = 0; i < l-1; i++){
        	int tmp = S.top();
        	S.pop();
        	Stack.push(tmp);
        }
    }

    // Get the front element.
    int peek(void) {
        int l = Stack.size();
        stack<int> S;
        for(int i = 0; i < l; i ++) {
        	int tmp = Stack.top();
        	Stack.pop();
        	S.push(tmp);
        }
        int ret = S.top();
        for(int i = 0; i < l; i++){
        	int tmp = S.top();
        	S.pop();
        	Stack.push(tmp);
        }
        return ret;
    }

    // Return whether the queue is empty.
    bool empty(void) {
        return Stack.empty();
    }
};
